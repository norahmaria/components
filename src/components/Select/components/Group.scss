@use '../../../scss/variables' as *;

.select-nm > .options-nm > .group-nm + .group-nm,
.select-nm > .options-nm > .group-nm ~ .group-nm {
  ul {
    border-bottom: none;
    margin-bottom: 0;
    padding-bottom: 0;
  }
}

.select-nm > .options-nm > .group-nm:first-of-type {
  ul {
    padding-top: 1.25rem;
  }
}

.select-nm > .options-nn .option-nm + .group-nm {
  margin-top: 0.75rem;
}

.select-nm > .options-nm > .group-nm {
  ul {
    display: grid;
    gap: 0.35rem;
    width: 100%;

    .title-nm {
      display: flex;
      gap: 0.3rem;
      align-items: center;

      font-size: 0.875em;
      padding: 0 0.5em 0.5em 0.5rem;
      overflow: hidden;

      text-transform: uppercase;
      font-weight: 600;

      color: inherit;
      color: rgba($black, 0.65);

      svg {
        transform: rotate(180deg);

        path,
        rect {
          fill: rgba($black, 0.45);
        }
      }
    }

    .option-nm {
      &:last-of-type {
        padding-bottom: 0.25rem;
      }
    }
  }
}
