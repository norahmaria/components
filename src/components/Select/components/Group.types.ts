import React from 'react'

interface GroupProps {
  title: string
  children: React.ReactChild | React.ReactChild[]
}

export default GroupProps
