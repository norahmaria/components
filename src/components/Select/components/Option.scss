@use '../../../scss/variables' as *;

.select-nm > .options > .option,
.select-nm > .options > .group .option {
  display: flex;
  justify-content: space-between;
  align-items: center;
  list-style: none;
  gap: 0.5rem;

  border: 1px solid transparent;
  border-radius: $br;

  padding: 0.4em 0.5em;

  transition: background 0.25s ease-in-out, color 0.25s ease-in-out,
    border 0.25s ease-in-out, padding 0.25s ease-in-out;
  cursor: pointer;

  .label {
    width: 100%;
    display: flex;
    align-items: center;
    gap: 0.5rem;
  }

  &,
  &[aria-selected='true'] {
    &:hover,
    &:focus-visible,
    &:active {
      outline: none;
      background: darken($white, 5%);
      border: 1px solid lighten($black, 75.5%);
      padding-left: 1em;
    }

    &:focus-visible {
      background: lighten($black, 77.5%);
    }
  }

  &[aria-disabled='true'] {
    background: darken($white, 3%);
    color: #898a8a;
    cursor: not-allowed;

    &:hover {
      padding: 0.4em 0.5em;
    }

    svg {
      opacity: 0.45;
      filter: grayscale(100%);
    }
  }

  &[aria-selected='true'] {
    font-weight: 500;
  }

  &:first-child {
    margin-top: 0.9rem;
  }

  &:last-child {
    margin-bottom: 0.9rem;
  }
}

@each $name, $color in $colors {
  @if $name == 'neutral' {
    .select-nm.#{$name} > .options > .option,
    .select-nm.#{$name} > .options > .group .option {
      &[aria-selected='true'] {
        background: scale-color($color, $saturation: 5%, $lightness: 50%);
        border: 1px solid scale-color($color, $saturation: 5%, $lightness: 50%);
      }
    }
  } @else {
    .select-nm.#{$name} > .options > .option,
    .select-nm.#{$name} > .options > .group .option {
      &[aria-selected='true'] {
        background: scale-color($color, $saturation: 5%, $lightness: 90%);
        border: 1px solid scale-color($color, $saturation: 5%, $lightness: 90%);
      }
    }
  }
}
