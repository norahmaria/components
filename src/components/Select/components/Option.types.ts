interface OptionProps {
  value: string | number
  leftIcon?: JSX.Element | string
  rightIcon?: JSX.Element | string
  disabled?: boolean
  children: string
}

export default OptionProps
