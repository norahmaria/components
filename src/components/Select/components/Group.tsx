import React from 'react'
import './Group.scss'

import { ReactComponent as LineArrow } from '../../../assets/Arrow_Line.svg'
import GroupProps from '../components/Group.types'

const Group: React.FC<GroupProps> = ({ title, ...props }) => {
  return (
    <ul className="group" {...props}>
      <div className="title">
        {/* <LineArrow /> */}
        {title}
      </div>
      {props.children}
    </ul>
  )
}

export default Group
