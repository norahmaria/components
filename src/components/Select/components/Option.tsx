import React from 'react'
import './Option.scss'

import OptionProps from './Option.types'

const Option = ({}: OptionProps) => <></>

export default Option
