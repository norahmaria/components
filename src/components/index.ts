export { default as Button } from './Button/Button'
export { default as Checkbox } from './Checkbox/Checkbox'

import SelectComponent from './Select/Select'
import Option from './Select/components/Option'
import Group from './Select/components/Group'

import SelectProps from './Select/Select.types'
import OptionProps from './Select/components/Option.types'
import GroupProps from './Select/components/Group.types'

type SelectWithChildren = React.FC<SelectProps> & {
  Option: React.FC<OptionProps>
  Group: React.FC<GroupProps>
}

const Select: SelectWithChildren = SelectComponent as SelectWithChildren
Select.Option = Option
Select.Group = Group

export { Select }
