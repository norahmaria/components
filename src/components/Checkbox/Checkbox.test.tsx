import React from 'react'
import { act, render, waitFor } from '@testing-library/react'

import { Checkbox } from '../../components/index'

describe('Checkbox Component', () => {
  it('Can render', () => {
    const { getByTestId } = render(
      <Checkbox
        label="Checkbox"
        id="checkbox"
        onCheckboxChange={() => {}}
      />
    )

    expect(getByTestId('checkbox-wrapper')).toMatchSnapshot()
  })

  it('Can run the onCheckboxChange event on changes', () => {
    const onCheckboxChange = jest.fn()

    const { getByTestId } = render(
      <Checkbox
        label="Checkbox"
        id="checkbox"
        onCheckboxChange={onCheckboxChange}
      />
    )

    const label = getByTestId('checkbox-label')
    label.click()

    expect(onCheckboxChange).toHaveBeenCalled()
  })

  it('Does not run onCheckboxChange events if disabled', () => {
    const onCheckboxChange = jest.fn()

    const { getByTestId } = render(
      <Checkbox
        label="Checkbox"
        id="checkbox"
        disabled={true}
        onCheckboxChange={onCheckboxChange}
      />
    )

    const label = getByTestId('checkbox-label')
    label.click()

    expect(onCheckboxChange).not.toHaveBeenCalled()
  })

  it('Can toggle checked / unchecked', () => {
    const { getByTestId } = render(
      <Checkbox
        label="Checkbox"
        id="checkbox"
        onCheckboxChange={() => {}}
      />
    )

    const label = getByTestId('checkbox-label')
    label.click()

    const input = getByTestId('checkbox-input')
    expect(input).toBeChecked()
  })
})
