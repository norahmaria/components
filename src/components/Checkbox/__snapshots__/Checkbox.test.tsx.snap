// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component Can render 1`] = `
<div
  class="
        checkbox-nm 
        size-medium 
        color-primary 
        undefined
      "
  data-testid="checkbox-wrapper"
>
  <input
    data-testid="checkbox-input"
    id="checkbox"
    type="checkbox"
  />
  <label
    class="right"
    data-testid="checkbox-label"
    for="checkbox"
  >
    <svg />
    Checkbox
  </label>
</div>
`;
